File: 739_每日温度.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <cassert>
using namespace std;

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        /*
         * 单调栈法：时间 O(n)，空间 O(n)
         * 维护单调递减栈，存储温度的索引
         * 当遇到更高温度时，弹出栈顶并计算天数差
         */
        int n = temperatures.size();
        vector<int> answer(n, 0);
        stack<int> st;  // 单调递减栈，存储索引

        for (int i = 0; i < n; i++) {
            // 当栈不为空且当前温度 > 栈顶温度
            while (!st.empty() && temperatures[i] > temperatures[st.top()]) {
                int prevIndex = st.top();
                st.pop();
                answer[prevIndex] = i - prevIndex;
            }
            st.push(i);
        }

        return answer;
    }
};

// ==================== 测试代码 ====================

void printResult(const vector<int>& result) {
    cout << "[";
    for (size_t i = 0; i < result.size(); i++) {
        cout << result[i];
        if (i < result.size() - 1) cout << ", ";
    }
    cout << "]" << endl;
}

int main() {
    Solution solution;

    // 测试用例 1
    cout << "==================================================" << endl;
    cout << "测试用例 1:" << endl;
    vector<int> temps1 = {73, 74, 75, 71, 69, 72, 76, 73};
    cout << "温度: [73, 74, 75, 71, 69, 72, 76, 73]" << endl;
    vector<int> result1 = solution.dailyTemperatures(temps1);
    cout << "结果: "; printResult(result1);
    assert(result1.size() == 8 && result1[0] == 1 && result1[2] == 4);
    cout << "✓ 通过" << endl;

    // 测试用例 2
    cout << "\n==================================================" << endl;
    cout << "测试用例 2:" << endl;
    vector<int> temps2 = {30, 40, 50, 60};
    cout << "温度: [30, 40, 50, 60]" << endl;
    vector<int> result2 = solution.dailyTemperatures(temps2);
    cout << "结果: "; printResult(result2);
    assert(result2.size() == 4 && result2[0] == 1 && result2[3] == 0);
    cout << "✓ 通过" << endl;

    // 测试用例 3
    cout << "\n==================================================" << endl;
    cout << "测试用例 3:" << endl;
    vector<int> temps3 = {30, 60, 90};
    cout << "温度: [30, 60, 90]" << endl;
    vector<int> result3 = solution.dailyTemperatures(temps3);
    cout << "结果: "; printResult(result3);
    assert(result3.size() == 3 && result3[0] == 1 && result3[2] == 0);
    cout << "✓ 通过" << endl;

    // 测试用例 4: 递减温度
    cout << "\n==================================================" << endl;
    cout << "测试用例 4 (递减温度):" << endl;
    vector<int> temps4 = {90, 80, 70, 60};
    cout << "温度: [90, 80, 70, 60]" << endl;
    vector<int> result4 = solution.dailyTemperatures(temps4);
    cout << "结果: "; printResult(result4);
    assert(result4.size() == 4 && result4[0] == 0 && result4[3] == 0);
    cout << "✓ 通过" << endl;

    // 测试用例 5: 相同温度
    cout << "\n==================================================" << endl;
    cout << "测试用例 5 (相同温度):" << endl;
    vector<int> temps5 = {50, 50, 50};
    cout << "温度: [50, 50, 50]" << endl;
    vector<int> result5 = solution.dailyTemperatures(temps5);
    cout << "结果: "; printResult(result5);
    assert(result5.size() == 3 && result5[0] == 0 && result5[2] == 0);
    cout << "✓ 通过" << endl;

    cout << "\n==================================================" << endl;
    cout << "所有测试用例通过！" << endl;

    return 0;
}

File: 739_每日温度.py
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        """
        单调栈法：时间 O(n)，空间 O(n)
        维护单调递减栈，存储温度的索引
        当遇到更高温度时，弹出栈顶并计算天数差
        """
        n = len(temperatures)
        answer = [0] * n
        stack = []  # 单调递减栈，存储索引

        for i, temp in enumerate(temperatures):
            # 当栈不为空且当前温度 > 栈顶温度
            while stack and temp > temperatures[stack[-1]]:
                prev_index = stack.pop()
                answer[prev_index] = i - prev_index
            stack.append(i)

        return answer


# ==================== 测试代码 ====================

def test():
    solution = Solution()

    # 测试用例 1
    print("=" * 50)
    print("测试用例 1:")
    temps1 = [73, 74, 75, 71, 69, 72, 76, 73]
    print(f"温度: {temps1}")
    result1 = solution.dailyTemperatures(temps1)
    print(f"结果: {result1}")
    assert result1 == [1, 1, 4, 2, 1, 1, 0, 0], "测试用例 1 失败"
    print("✓ 通过")

    # 测试用例 2
    print("\n" + "=" * 50)
    print("测试用例 2:")
    temps2 = [30, 40, 50, 60]
    print(f"温度: {temps2}")
    result2 = solution.dailyTemperatures(temps2)
    print(f"结果: {result2}")
    assert result2 == [1, 1, 1, 0], "测试用例 2 失败"
    print("✓ 通过")

    # 测试用例 3
    print("\n" + "=" * 50)
    print("测试用例 3:")
    temps3 = [30, 60, 90]
    print(f"温度: {temps3}")
    result3 = solution.dailyTemperatures(temps3)
    print(f"结果: {result3}")
    assert result3 == [1, 1, 0], "测试用例 3 失败"
    print("✓ 通过")

    # 测试用例 4: 递减温度
    print("\n" + "=" * 50)
    print("测试用例 4 (递减温度):")
    temps4 = [90, 80, 70, 60]
    print(f"温度: {temps4}")
    result4 = solution.dailyTemperatures(temps4)
    print(f"结果: {result4}")
    assert result4 == [0, 0, 0, 0], "测试用例 4 失败"
    print("✓ 通过")

    # 测试用例 5: 相同温度
    print("\n" + "=" * 50)
    print("测试用例 5 (相同温度):")
    temps5 = [50, 50, 50]
    print(f"温度: {temps5}")
    result5 = solution.dailyTemperatures(temps5)
    print(f"结果: {result5}")
    assert result5 == [0, 0, 0], "测试用例 5 失败"
    print("✓ 通过")

    print("\n" + "=" * 50)
    print("所有测试用例通过！")


if __name__ == "__main__":
    test()
